mne/epochs.py

```diff
@@ -2064,6 +2064,280 @@ def _drop_log_stats(drop_log, ignore=('IGNORED',)):
+        create `~mne.Epochs`, that is, the number of rows of the generated
+        metadata will equal the number of events with the specified IDs.
+        Moreover, metadata will only be generated for the events listed here.
+    tmin : float
+        Start of the time interval for event extraction in seconds, relative to
+        time-locked event of the respective epoch. This will typically be the
+        same ``tmin`` you'll pass to `~mne.Epochs`, but it may also be
+        different to include more or fewer events in the metadata.
+    tmax : float
+        Start of the time interval for event extraction in seconds, relative to
+        time-locked event of the respective epoch. This will typically be the
+        same ``tmax`` you'll pass to `~mne.Epochs`, but it may also be
+        different to include more or fewer events in the metadata.
+    sfreq : float
+        The sampling frequency during data acquisiton.
+    time_locked_events : list of str | str | None
+        This can be a subset of event names (keys of ``event_id``). Metadata
+        rows will only be created for the specified event(s), instead of for
+        all events present in ``event_id``. If ``None`` (default), create
+        one row for each event in ``event_id``.
+    keep_first : str | list of str | None
+        Specify hierarchical event descriptors(s) (HED) of events from which
+        only the **first occurrence only** within any time period shall be
+        kept, in case there are multiple. For example, you might have two
+        response events, ``response/left`` and ``response/right``; and in
+        trials with more than one response, you want to keep only the very
+        first response. In this case, you can set ``keep_first=['response']``.
+        This will add two new columns to the metadata: , ``response`` and
+        ``response_time``, with the time relative to the time-locked event.
+        If ``None`` all events are kept.
+    keep_last : list of str | None
+        Same as ``keep_first``, but for keeping the only the **last**
+        occurrence of matching events.
+
+    Returns
+    -------
+    metadata : pandas.DataFrame
+        The pre-assembled metadata. It contains the following columns:
+
+        - ``event_name``, with strings indicating the name of the time-locked
+          event for that specific time period
+
+        - one column per key in ``event_id``, with the same name; boolean
+          values indicating whether that specific event occurred within the
+          given time period
+
+        - if applicable, additional columns for ``keep_first`` and
+          ``keep_last`` event names, if HEDs were specified; the values will
+          be strings indicating which event was matched through the HED
+          pattern
+
+        - columns named ``[event_name]_time``, whose values indicate the time
+          of event occurrence in seconds, relative to the time-locked event
```

Comment:
I think a digested version of this info ought to be in the introductory description.  i.e., something like

> This function mimics the epoching process (constructs time windows around events of interest) and collates information about other events that occurred within those time windows. The information is returned as a pandas dataframe suitable for use as epochs metadata: one row per time-locked event, and columns indicating presence/absence and latency of each ancillary event type.

Reply:
writing that makes me wonder why the presence/absence column is needed; couldn't `np.nan` in the latency column be enough information to indicate absence?

Reply:
> writing that makes me wonder why the presence/absence column is needed; couldn't `np.nan` in the latency column be enough information to indicate absence?

This is a fair point and I think eventually it's a bit a question of preference.

- To me, it's foremost a semantic consideration: The names of the boolean presence/absence columns are the exact same as the event names (i.e., those in `event_id` or `keep_first`/`keep_last`). I wouldn't feel comfy putting latencies in there as it wouldn't be clear that these **are** latencies, at least not without using multiindex columns, which are not supported by Epochs metadata.

- A second argument pertains to indexing. Look at this example:

```python
> print(epochs.metadata[['response/right', 'response/right_time']].head())

   response/right  response/right_time
0           False                  NaN
1           False                  NaN
2            True             0.508789
3           False                  NaN
4           False                  NaN
```

Now, assume we want to use `DataFrame.query()` to retrieve all rows where we have a `response/right` event. If `response/right` is a boolean, it's straightforward:

```python
> print(epochs.metadata
        .query('`response/right`')
        [['response/right', 'response/right_time']]
        .head())

    response/right  response/right_time
2             True             0.508789
8             True             0.484375
10            True             0.473633
12            True             0.458984
13            True             0.401367
```

If, however, the only way to figure whether an event occurred is by checking whether its value (latency) is `NaN`, things get messy, as there is no simple way to do this check using the `query()` method. Instead, one can abuse the fact that `NaN` doesn't equal itself:
```python
> print(epochs.metadata
        .query('`response/right_time` == `response/right_time`')
        [['response/right', 'response/right_time']]
        .head())

     response/right  response/right_time
2              True             0.508789
8              True             0.484375
10             True             0.473633
12             True             0.458984
13             True             0.401367
```
which works, but is pretty ugly.

Reply:
> I think a digested version of this info ought to be in the introductory description. i.e., something like
> 
> > This function mimics the epoching process (constructs time windows around events of interest) and collates information about other events that occurred within those time windows. The information is returned as a pandas dataframe suitable for use as epochs metadata: one row per time-locked event, and columns indicating presence/absence and latency of each ancillary event type.

I've adopted this proposal now :)